File: ios/Wellvo/Services/AnalyticsService.swift
import Foundation

/// Privacy-first analytics using TelemetryDeck.
/// No personal data is collected — only anonymous usage signals.
/// TelemetryDeck is GDPR-compliant and does not use cookies or fingerprinting.
///
/// To integrate:
/// 1. Add TelemetryDeck SPM package: https://github.com/TelemetryDeck/SwiftSDK
/// 2. Set TELEMETRY_APP_ID in BuildConfig.xcconfig
/// 3. Initialize in WellvoApp.init()
///
/// Usage: AnalyticsService.shared.track(.checkIn)
actor AnalyticsService {
    static let shared = AnalyticsService()

    private var isInitialized = false

    // MARK: - Event Names

    enum Event: String {
        // Auth
        case signIn = "auth.sign_in"
        case signUp = "auth.sign_up"
        case signOut = "auth.sign_out"

        // Check-in
        case checkIn = "checkin.completed"
        case checkInOffline = "checkin.offline_queued"
        case checkInOfflineSynced = "checkin.offline_synced"
        case moodSubmitted = "checkin.mood_submitted"

        // Owner actions
        case dashboardViewed = "owner.dashboard_viewed"
        case onDemandCheckInSent = "owner.on_demand_sent"
        case receiverInvited = "owner.receiver_invited"
        case receiverRemoved = "owner.receiver_removed"
        case settingsSaved = "owner.settings_saved"
        case reportExported = "owner.report_exported"
        case historyViewed = "owner.history_viewed"

        // Subscription
        case subscriptionStarted = "subscription.started"
        case subscriptionRestored = "subscription.restored"
        case subscriptionViewOpened = "subscription.view_opened"

        // Onboarding
        case onboardingStarted = "onboarding.started"
        case onboardingCompleted = "onboarding.completed"
        case onboardingStepViewed = "onboarding.step_viewed"

        // Notifications
        case notificationPermissionGranted = "notification.permission_granted"
        case notificationPermissionDenied = "notification.permission_denied"

        // Alerts
        case alertViewed = "alert.viewed"
        case alertDismissed = "alert.dismissed"

        // App lifecycle
        case appOpened = "app.opened"
        case appBackgrounded = "app.backgrounded"
    }

    // MARK: - Initialization

    /// Initialize TelemetryDeck with the app ID from build config.
    /// Call this once in WellvoApp.init().
    func initialize() {
        #if !DEBUG
        // Only initialize in release builds
        // TelemetryDeck.initialize(config: .init(appID: Configuration.telemetryAppID))
        isInitialized = true
        #endif
    }

    // MARK: - Tracking

    /// Track an analytics event with optional metadata.
    /// All data is anonymous — no user IDs or personal info.
    func track(_ event: Event, properties: [String: String] = [:]) {
        guard isInitialized else { return }

        // TelemetryDeck.signal(event.rawValue, parameters: properties)
        // Uncomment the line above once TelemetryDeck SPM package is added.

        #if DEBUG
        print("[Analytics] \(event.rawValue) \(properties.isEmpty ? "" : properties.description)")
        #endif
    }

    /// Track a screen view.
    func trackScreen(_ name: String) {
        track(.dashboardViewed, properties: ["screen": name])
    }
}

File: android/app/src/main/java/net/wellvo/android/services/AnalyticsService.kt
package net.wellvo.android.services

import android.app.Application
import android.util.Log
import com.telemetrydeck.sdk.TelemetryDeck
import dagger.hilt.android.qualifiers.ApplicationContext
import net.wellvo.android.BuildConfig
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AnalyticsService @Inject constructor(
    @ApplicationContext private val context: android.content.Context
) {
    companion object {
        private const val TAG = "AnalyticsService"

        // Auth events
        const val SIGN_IN = "sign_in"
        const val SIGN_UP = "sign_up"
        const val SIGN_OUT = "sign_out"

        // Check-in events
        const val CHECK_IN_COMPLETED = "check_in_completed"
        const val CHECK_IN_OFFLINE_QUEUED = "check_in_offline_queued"
        const val CHECK_IN_OFFLINE_SYNCED = "check_in_offline_synced"
        const val MOOD_SUBMITTED = "mood_submitted"

        // Dashboard events
        const val DASHBOARD_VIEWED = "dashboard_viewed"
        const val ON_DEMAND_SENT = "on_demand_sent"

        // Family events
        const val RECEIVER_INVITED = "receiver_invited"
        const val RECEIVER_REMOVED = "receiver_removed"

        // Settings events
        const val SETTINGS_SAVED = "settings_saved"
        const val REPORT_EXPORTED = "report_exported"

        // History events
        const val HISTORY_VIEWED = "history_viewed"

        // Subscription events
        const val SUBSCRIPTION_STARTED = "subscription_started"
        const val SUBSCRIPTION_RESTORED = "subscription_restored"
        const val SUBSCRIPTION_VIEW_OPENED = "subscription_view_opened"

        // Onboarding events
        const val ONBOARDING_STARTED = "onboarding_started"
        const val ONBOARDING_COMPLETED = "onboarding_completed"
        const val ONBOARDING_STEP_VIEWED = "onboarding_step_viewed"

        // Notification events
        const val NOTIFICATION_PERMISSION_GRANTED = "notification_permission_granted"
        const val NOTIFICATION_PERMISSION_DENIED = "notification_permission_denied"

        // Alert events
        const val ALERT_VIEWED = "alert_viewed"
        const val ALERT_DISMISSED = "alert_dismissed"

        // App lifecycle events
        const val APP_OPENED = "app_opened"
        const val APP_BACKGROUNDED = "app_backgrounded"
    }

    private var isInitialized = false

    fun initialize() {
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "Analytics disabled in debug builds")
            return
        }

        val appId = BuildConfig.TELEMETRY_DECK_APP_ID
        if (appId.isBlank()) {
            Log.w(TAG, "TelemetryDeck app ID not configured")
            return
        }

        try {
            val builder = TelemetryDeck.Builder()
                .appID(appId)

            if (context is Application) {
                builder.build(context)
            }
            isInitialized = true
            Log.d(TAG, "TelemetryDeck initialized")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize TelemetryDeck", e)
        }
    }

    fun track(event: String, properties: Map<String, String> = emptyMap()) {
        if (!isInitialized) return

        try {
            if (properties.isEmpty()) {
                TelemetryDeck.signal(event)
            } else {
                TelemetryDeck.signal(event, properties)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to track event: $event", e)
        }
    }

    fun trackScreen(screenName: String) {
        track("screen_viewed", mapOf("screen" to screenName))
    }
}
